pymeasure/adapters/adapter.py

```diff
@@ -33,48 +37,135 @@ class Adapter:
 
     This class should only be inherited from.
 
-    :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+    :param preprocess_reply: (deprecated) optional callable used to preprocess
+        strings received from the instrument. The callable returns the
+        processed string.
+    :param float query_delay: Time in s to wait after writing and before reading.
+    :param log: Parent logger of the 'Adapter' logger.
     :param kwargs: all other keyword arguments are ignored.
     """
 
-    def __init__(self, preprocess_reply=None, **kwargs):
+    def __init__(self, preprocess_reply=None, query_delay=0, log=None, **kwargs):
         self.preprocess_reply = preprocess_reply
+        self.query_delay = query_delay
         self.connection = None
+        if log is None:
+            self.log = logging.getLogger("Adapter")
+        else:
+            self.log = log.getChild("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn("Deprecated in Adapter, pass it to Instrument.__init__, instead.",
+                 FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
+        self.close()
+
+    def close(self):
+        """Close the connection."""
         if self.connection is not None:
             self.connection.close()
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
+    def write(self, command, **kwargs):
```

Comment:
I like the interface of `Adapter` with all the problematic methods now deprecated. The methods defined do not have call between themselves.
I see lack of `write_binary_values` which is the opposite of `binary_values` and it is very useful when large data has to be transmitted or received from instrument.

Reply:
I had a lengthy discussion regarding reading / writing binary data.
The result (not yet implemented) was, that the instrument has read /write binary values, as it also implements the normal values.
In order to reduce maintenence overhead and simultaneously have powerful methods, I plan to use the útil class from pyvisa for the formatting/decoding.

That way it works the same independent from the adapter (visa or serial...).